### PR TITLE
Get rid of tsd

### PR DESCRIPTION
**Pull request by markw65**
Get rid of `tsd` by rewriting `expectType` in a way that lets vanilla typescript check the types with the same precision.

### CODE REVIEW

File: test/types/peg.test-d.ts [hunk 1]
```diff
@@ -16,6 +15,15 @@ const src = readFileSync(
...
 
+function expectExact<T>() {
+  return function<U extends T>(
+    tValue: U
+  ): [T] extends [U] ? () => U : ["Argument of type", U, "does not match", T] {
+    // @ts-expect-error Just for type checking
+    return () => tValue;
+  };
+}
+
```
Comment (markw65) [inline test/types/peg.test-d.ts:26]: On the one hand, I was inclined to keep the `expectType` name, but on the other, that would cause confusion for people familiar with `tsd`. But we could go either way.

We could also break this into two functions, which might make it easier to spot that you accidentally omitted the trailing `()`:

```
function exact<T>() {
   return function<U extends T>(
     tValue: U
   ): [T] extends [U] ? () => U : [U, "does not match", T] {
     // @ts-expect-error Just for type checking
     return () => tValue;
   };
 }
function check<T>(fn: ()=>T): void {
  fn();
}
```

And then call `check(exact<T>()(arg))`.

In fact, this makes the error message work better. Now its:
```
Argument of type '[string, "does not match", string | number]' is not assignable to parameter of type '() => unknown'.
  Type '[string, "does not match", string | number]' provides no match for the signature '(): unknown'.
```

**Comment by hildjj** [comment]
This one is next after #430

**Comment by hildjj** [comment]
Rebase and changelog, please.

**Comment by markw65** [comment]
Updated. I'll just note one other way to do this that might be a bit clearer. 
See [typescript playground](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBZJcWD1Rrg9aQKkAUVWDCpAB4cAA+AAUAEpQABvRQlJyofFeCnmamcVn3WkKHjCTlcoUlNAANQY0HYkHyDxVPN1AG0cABdOUo7jCQ33M0AflAvNAAF4OQ9QEaAEQVMjMdjQpKpVDtD38e78D3ceD6fqDUCIOoaEPeM2C1WgFQAAVQglKEPpqFzZDY2IAUuxBDRkvBsUGiCZsLB4sMVcLIKKyF5HS71ZrtQBuFUAXwHg+CCE4FdAoKdoA9VfgHrEwjn8AXAB8PdJHB6B+PJ3hnbP54uGMvjxut2QPReGAAvHfBPOM5lslcL0Br2eXj3cnlc1Z-jyA5PqgTIsqgrJvoun6btuv5cnggHAXSz7gZBx4fl+cG8lyUFIeSlJSlc7DSLWkAAPLJOyv4CiqIpijihENNKsprPKlreNyLY9lqOp6iUBqgMaZpsRaVo2qA9r8kIAD6azgfkqBkNqg7ukJXo+n6KI0ESZFJmGs6RtGAw0PGqCJvwpq0WmmbZrmKEFt8xagGWk5VjW7T1vSTacMw3H0R2AqyfJVKKcpkDDkOI7BIIJFkZR6GrtBWFXvBAEBMFL6oAOsWke0CVQZhsGpThiEZYIclZTlcX5VRhUwd+8F4eVlXgdVeVEAVGENdhf54QOQA)

Basically, instead of exactType<T>()(x)(), we could use `subtypeOf<T>()(x).is_exact`. It's the same trick, but (in my opinion) it reads a bit better. Without the is_exact, it just checks for any subtype; with the is_exact it checks for exact types (the original exactType<T> does the same if you omit the final `()` - but its less clear to readers what the `()` is there for)

**Comment by hildjj** [comment]
I don't have a strong opinion, so I'll defer to your taste.  Feel free to send another PR if you decide you like another way better.

**Comment by reverofevil** [comment]
Well, either trick is using heavily unstable features of TypeScript.

I wouldn't recommend either approach for future rewrites, but for a small localized change to remove `tsd` this is already good.

**Comment by markw65** [comment]
> Well, either trick is using heavily unstable features of TypeScript

I checked it back to ts-3.3.3. Same behavior everywhere - so absent some statement of intent to change the behavior (perhaps you've seen something?) I think it's pretty stable. Its just currying; its hard to see how it could break without breaking huge amounts of typescript...

**Comment by reverofevil** [comment]
Sorry, I didn't mean the problem is with currying. Problem is with conditional types, and, more precisely, them lacking any foundation in type theory. They might look like they're working, but even that equality check is incorrect. There're ["better" type equalities](https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-421529650) in TS, but no single correct one. Unfortunately even type equality is still an open problem.

Mapped object types, conditional types, indexed types, type guards and, well, majority of TS's type system is unsound and inconsistent, and it's best to avoid any of that if reliability is at stake.
